src/hash_table/chaining.rs

```diff
@@ -0,0 +1,1628 @@
+// Copyright 2019 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+//! Intrusive hash table that uses "separate chaining" for collision resolution.
+#[cfg(any(test, feature = "std"))]
+use std::collections::hash_map::RandomState;
+
+use core::borrow::Borrow;
+use core::cmp;
+use core::hash::{BuildHasher, Hash, Hasher};
+use core::ops::Deref;
+use core::iter;
+use core::fmt;
+
+use crate::adapter::Adapter;
+use crate::array::Array;
+use crate::bits::{Bits, BitsArray};
+use crate::dynamic_array::DynamicArray;
+use crate::hash_adapter::HashAdapter;
+pub use crate::hash_table::ChainedLoadFactor;
+use crate::key_adapter::KeyAdapter;
+use crate::linked_list::Link as LinkedListLink;
+use crate::linked_list::LinkedList;
+//use crate::intrusive_pointer::IntrusivePointer;
+use crate::hash_table::DEFAULT_CHAINED_LOAD_FACTOR;
+
+#[cfg(any(test, feature = "std"))]
+/// An intrusive hash table.
+///
+/// When this collection is dropped, all elements linked into it will be
+/// converted back to owned pointers and dropped.
+#[derive(Clone)]
+pub struct ChainedHashTable<A, B, C, S = std::collections::hash_map::RandomState>
+where
+    A: Adapter<Link = LinkedListLink>,
+    B: Array<Item = LinkedList<A>>,
+    C: Array,
+{
+    adapter: A,
+    buckets: B,
+    buckets_bits: C,
+    buckets_len: usize,
+    hash_builder: S,
```

Comment:
As I mentioned above, it is probably better to leave the choice of hasher to the adapter. As a bonus this eliminates our dependency on `std`.